File: labview/src/server/cache.ts
/**
 * The scan cache, and the one rule that makes an explicit rescan trustworthy.
 *
 * A scan of a real fleet takes seconds — one Engine sweep plus two API exchanges — and
 * several callers can want a result inside that window: the warm scan at boot, a page
 * load, a timer expiring, an operator pressing Rescan. Sharing one build between them is
 * the point of this module, and it is also where a rescan can quietly lie.
 *
 * A build reads the compose files once, at its start. So a build that began *before* the
 * operator edited a file cannot contain that edit, and handing it to a rescan that
 * arrived *after* the edit answers with the pre-edit fleet — which looks exactly like
 * "LabView did not re-read the files". Hence the rule: a forced request may only be
 * answered by a build that started after the request arrived. Everything else here is
 * ordinary coalescing.
 *
 * Separated from `startServer` because the race is the behaviour: with `build` and `now`
 * injected, every ordering is assertable without a listening socket or a real clock.
 */

export interface ScanCacheDeps<T, R = void> {
  /**
   * Produce a fresh value. Called at most once per concurrent wave of requests, and given
   * the request that *started* this build — see {@link ScanCache.get} for who that is.
   */
  build: (req: R) => Promise<T>;
  /** How long a value may be served without rebuilding. `0` disables reuse. */
  ttlMs: number;
  /** Injected for determinism in tests. Defaults to `Date.now`. */
  now?: () => number;
  /**
   * Called once per completed build — not once per caller waiting on it.
   *
   * `prev` is the value this build replaced, or `undefined` for the first build, which is
   * what lets a caller report what changed. `forced` distinguishes an explicit rescan
   * from a timer expiry, because the two deserve different noise levels.
   */
  onBuilt?: (next: T, prev: T | undefined, info: { forced: boolean }) => void;
}

export interface ScanCache<T, R = void> {
  /**
   * The current value, rebuilding when it is stale or when `force` is set.
   *
   * Rejects when the build rejects; the previously cached value is left intact, so the
   * next caller reads the old value or triggers a fresh attempt rather than inheriting
   * the failure.
   *
   * `req` is handed to `build` **only if this call is the one that starts a build.** A
   * caller that joins an in-flight build gets that build's value, decided by whoever
   * started it, and its own `req` is discarded — which is the existing coalescing rule and
   * not a new wrinkle. It is stated here because `req` now carries a behaviour choice
   * rather than nothing at all, so the value returned may not be the one this caller asked
   * for. What keeps that honest is that the result says which was used: `ScanMeta.probe`
   * reports what the build that produced it actually did.
   *
   * Passing it per call rather than holding it on the cache is the point. Shared mutable
   * state would let a passive TTL rebuild pick up a caller's override, or two callers
   * overwrite each other's — and either way the value that ran would be nobody's request.
   */
  get(force: boolean, req: R): Promise<T>;
  /** The cached value without ever building. For callers that only want what is already known. */
  peek(): T | undefined;
}

export function createScanCache<T, R = void>(deps: ScanCacheDeps<T, R>): ScanCache<T, R> {
  const now = deps.now ?? Date.now;
  let value: T | undefined;
  let builtAt = 0;
  let inflight: Promise<T> | null = null;
  let inflightStartedAt = 0;

  async function get(force: boolean, req: R): Promise<T> {
    const requestedAt = now();
    if (!force && value !== undefined && now() - builtAt < deps.ttlMs) return value;

    // A loop, not an `if`: after awaiting one build another caller may have started the
    // next, and this request has to re-ask the same question of it.
    for (;;) {
      const current = inflight;
      if (!current) break;
      // A passive request takes whatever is being built. A forced one only accepts a
      // build that began after it asked — otherwise it would return a scan of the files
      // as they were before the edit that prompted the click.
      if (!force || inflightStartedAt >= requestedAt) return current;
      // Wait it out rather than starting a second build alongside it: two concurrent
      // sweeps would double the load on the socket proxy to produce one answer. The
      // failure is swallowed here only so this request goes on to build for itself —
      // the original caller still sees the rejection.
      await current.catch(() => undefined);
    }

    inflightStartedAt = now();
    const build = deps
      .build(req)
      .then((next) => {
        const prev = value;
        value = next;
        builtAt = now();
        deps.onBuilt?.(next, prev, { forced: force });
        return next;
      })
      .finally(() => {
        if (inflight === build) inflight = null;
      });
    inflight = build;
    return build;
  }

  return { get, peek: () => value };
}

File: labview/internal/cache/cache.go
// Package cache is §17's build cache: one scan shared by every reader, and the rule that decides
// when a new one has to be run.
//
// **A forced request may only be answered by a build that started after it arrived.** That is the
// whole contract, and every line below exists to keep it. A caller who pressed Rescan is asking
// about the fleet as it is now; handing them a build that began before they asked would answer a
// question they did not put. The TTL, the coalescing and the once-per-build callback all follow from
// that one sentence.
//
// There are at most two builds at any moment: the one running, and the one queued behind it. A
// forced request that the running build cannot answer does not start a second build alongside it —
// it queues, and every later request that also cannot use the running build queues with it. That is
// what makes two forced requests arriving together share one build rather than run two scans of the
// same fleet.
//
// The clock and the build function are injected, so all five of §17's consequences are testable
// without waiting.
package cache

import (
	"context"
	"sync"
	"time"

	"github.com/nrosier/labview/internal/changes"
	"github.com/nrosier/labview/internal/payload"
)

// Build is one scan. It is handed the probe override this build was asked for, and it never fails —
// a scan that could not read anything still produces a payload saying so (I4).
type Build func(ctx context.Context, probe *bool) payload.Overview

// Built is called once for each completed build, with the note describing what moved since the
// previous one and whether this build was forced.
//
// **Once per build, not once per waiter** (§17): ten browsers sharing one build produce one line in
// the log, because the log records what LabView did and it did one scan.
//
// It runs after the result is published, so a listener that reads the cache sees this build rather
// than the one before it. It must not call back into the cache. It is the one place in this package
// that may log — I7 keeps the analysis silent, not the server.
type Built func(note changes.Note, forced bool)

// Cache holds the current build and coordinates the next one.
type Cache struct {
	build Build
	now   func() time.Time
	ttl   time.Duration
	built Built

	mu sync.Mutex

	// current is the newest finished build, at is when the build that produced it *started*, and
	// probe is the override it was asked with.
	//
	// Started, not finished: the freshness a reader cares about is when LabView looked, and §17's
	// rule is stated in terms of when a build began. It also means a slow build does not get to
	// claim the whole TTL for a fleet it read minutes ago.
	current *payload.Overview
	at      time.Time
	probe   *bool

	// running is the build in progress and queued are the builds waiting to start, oldest first. A
	// waiter takes a pointer, releases the lock and blocks on that build's done channel, so the
	// lock is never held across a scan.
	//
	// The queue holds more than one only when two waiting requests disagree about the probe: every
	// request that can share a queued build does, so the ordinary case is a queue of one.
	running *build
	queued  []*build
}

// build is one run and everything a waiter needs in order to decide whether it may be answered by
// it.
type build struct {
	// startedAt is the whole of the forced-request rule: a forced request that arrived at T may
	// only be answered by a build whose startedAt is at or after T. Zero until the build starts,
	// which is why a queued build is joinable by anyone — it has not started yet, so it cannot
	// have started too early.
	startedAt time.Time

	// forced is whether *any* of this build's waiters forced it. A timer rebuild that a Rescan
	// joined did happen because somebody asked, and §17's cadence turns on whether anybody did.
	forced bool

	// probe is the override this build runs with (§13.7). The request that created the build owns
	// it; see probeSuits for who may join.
	probe *bool

	done   chan struct{}
	result payload.Overview
}

// Options configures a cache.
type Options struct {
	// Build is required.
	Build Build

	// TTL is how long a build stays fresh. Zero or negative means every request rebuilds, which is
	// a setting and not an error — somebody actively editing a fleet may want exactly that.
	TTL time.Duration

	// Now is the injected clock (§17, I7). Nil is time.Now.
	Now func() time.Time

	// Built is optional; nil means nobody is listening.
	Built Built
}

// New makes a cache. It runs nothing: the first build happens on the first Get, or on Warm.
func New(o Options) *Cache {
	now := o.Now
	if now == nil {
		now = time.Now
	}
	return &Cache{build: o.Build, now: now, ttl: o.TTL, built: o.Built}
}

// Get returns a build, running one if this request may not be answered by what is held.
func (c *Cache) Get(ctx context.Context, forced bool, probe *bool) payload.Overview {
	// The arrival time is read first, before the mutex, because it is what the whole rule is stated
	// against. Reading it after acquiring the lock would let a build that started while this call
	// was queued for the lock count as having started after the request arrived.
	arrived := c.now()

	c.mu.Lock()

	// A queued build has not started, so it necessarily starts after this request arrived — which is
	// what lets two forced requests arriving together share one build. Checked before the running
	// build so a forced request joins the pending scan rather than being measured against the stale
	// one.
	for _, b := range c.queued {
		if !probeSuits(probe, b.probe) {
			continue
		}
		if forced {
			b.forced = true
		}
		c.mu.Unlock()
		return c.wait(ctx, b)
	}

	if b := c.running; b != nil && probeSuits(probe, b.probe) {
		// A non-forced request joins whatever is running. A forced one may join only if that build
		// started at or after it arrived.
		if !forced || !b.startedAt.Before(arrived) {
			if forced {
				b.forced = true
			}
			c.mu.Unlock()
			return c.wait(ctx, b)
		}
	}

	if !forced && c.fresh(arrived) && probeSuits(probe, c.probe) {
		out := *c.current
		c.mu.Unlock()
		return out
	}

	b := &build{forced: forced, probe: probe, done: make(chan struct{})}

	// Queue behind the running build rather than beside it. The running build's waiters asked
	// earlier and are entitled to it, and two scans of one fleet at once would read the same Docker
	// socket twice for no gain.
	if c.running != nil {
		c.queued = append(c.queued, b)
		c.mu.Unlock()
		return c.wait(ctx, b)
	}

	b.startedAt = arrived
	c.running = b
	c.mu.Unlock()

	// The build runs on its own goroutine, with a context that cannot be cancelled by the request
	// that happened to start it: other waiters are owed the result, and a browser closing a tab is
	// not a reason to abandon a scan (I4). It is also what lets *this* caller be cancelled — a
	// caller that ran the build inline could not return until it finished, whatever its context
	// said. That goroutine also runs whatever queues behind this build, so a queued build has a
	// runner without the cache keeping one of its own.
	go c.drain(context.WithoutCancel(ctx), b)
	return c.wait(ctx, b)
}

// Warm runs a build in the background, for §18's *the cache MUST be warmed in the background at
// startup*. It returns immediately; a request arriving during the warm build joins it.
func (c *Cache) Warm(ctx context.Context) {
	go c.Get(ctx, false, nil)
}

// Peek returns the held build without running one, or nil. It is for a caller that wants to render
// what is already known — never for one that wants a scan.
func (c *Cache) Peek() *payload.Overview {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.current == nil {
		return nil
	}
	out := *c.current
	return &out
}

// fresh reports whether the held build may answer a non-forced request. The caller holds the lock.
func (c *Cache) fresh(at time.Time) bool {
	return c.current != nil && c.ttl > 0 && at.Sub(c.at) < c.ttl
}

// probeSuits reports whether a request asking for want may be answered by a build running with has
// (§13.7).
//
// A nil override is the caller saying *whatever configuration says* — no requirement of their own,
// so any build answers it, and the ordinary overview stays cheap after somebody has forced one
// probing rescan. An explicit override is a requirement about this build, so it is only answered by
// a build that ran with it: a caller who asked for the probe and got a payload without one would
// have been told nothing about the thing they asked about.
func probeSuits(want, has *bool) bool {
	return want == nil || (has != nil && *want == *has)
}

// drain runs b, then whatever queued behind it, until nothing is waiting. The context is already
// detached from the request that started it; cancellation reaches waiters through wait, never
// through the build.
func (c *Cache) drain(ctx context.Context, b *build) {
	for b != nil {
		out := c.build(ctx, b.probe)

		c.mu.Lock()
		previous := c.current
		b.result = out
		// The TTL is stamped from this build's own start, by the same arithmetic every build uses —
		// which is how a forced build resets the TTL (§17) without a rule of its own.
		c.current, c.at, c.probe = &out, b.startedAt, b.probe
		forced := b.forced

		// Take the oldest queued build, in arrival order: a request that waited longer is not made
		// to wait again behind one that arrived after it.
		var next *build
		if len(c.queued) > 0 {
			next, c.queued = c.queued[0], c.queued[1:]
			next.startedAt = c.now()
		}
		c.running = next
		c.mu.Unlock()

		close(b.done)

		if c.built != nil {
			c.built(changes.Describe(previous, out), forced)
		}

		b = next
	}
}

// wait blocks until the build finishes, or until the caller's context is done.
//
// A cancelled caller gets the previous build if there is one and an empty payload if there is not.
// The build itself carries on for the waiters that are still owed it.
func (c *Cache) wait(ctx context.Context, b *build) payload.Overview {
	select {
	case <-b.done:
		return b.result
	case <-ctx.Done():
		if held := c.Peek(); held != nil {
			return *held
		}
		return payload.Overview{}
	}
}
